File: go/internal/slhdsa/fors.go
package slhdsa

// FORS (Forest of Random Subsets) implementation per FIPS 205.

// FORSSKGen generates a FORS secret key value.
// FIPS 205 Algorithm 12.
func FORSSKGen(hs HashSuite, skSeed, pkSeed []byte, adrs *Address, idx int, n int) []byte {
	skAdrs := adrs.Copy()
	skAdrs.SetType(AddrForsPRF)
	skAdrs.SetKeyPairAddress(adrs.GetKeyPairAddress())
	skAdrs.SetTreeIndex(uint32(idx))
	return hs.PRF(pkSeed, skSeed, &skAdrs, n)
}

// FORSNode computes a node in a FORS tree.
// FIPS 205 Algorithm 13: fors_node(SK.seed, i, z, PK.seed, ADRS).
func FORSNode(hs HashSuite, skSeed, pkSeed []byte, i, z int, adrs *Address, p *Params) []byte {
	n := p.N

	if z == 0 {
		sk := FORSSKGen(hs, skSeed, pkSeed, adrs, i, n)
		adrs.SetType(AddrForsTree)
		adrs.SetTreeHeight(0)
		adrs.SetTreeIndex(uint32(i))
		return hs.F(pkSeed, adrs, sk, n)
	}

	left := FORSNode(hs, skSeed, pkSeed, 2*i, z-1, adrs, p)
	right := FORSNode(hs, skSeed, pkSeed, 2*i+1, z-1, adrs, p)

	adrs.SetType(AddrForsTree)
	adrs.SetTreeHeight(uint32(z))
	adrs.SetTreeIndex(uint32(i))

	concat := make([]byte, 0, 2*n)
	concat = append(concat, left...)
	concat = append(concat, right...)
	return hs.H(pkSeed, adrs, concat, n)
}

// FORSSign generates a FORS signature.
// FIPS 205 Algorithm 14: fors_sign(md, SK.seed, PK.seed, ADRS).
func FORSSign(hs HashSuite, md, skSeed, pkSeed []byte, adrs *Address, p *Params) []byte {
	n := p.N
	k := p.K
	a := p.A

	indices := base2b(md, a, k)

	sig := make([]byte, 0, k*(1+a)*n)

	for i := 0; i < k; i++ {
		idx := indices[i]

		// Secret key value
		sk := FORSSKGen(hs, skSeed, pkSeed, adrs, i*(1<<uint(a))+idx, n)
		sig = append(sig, sk...)

		// Authentication path
		for j := 0; j < a; j++ {
			s := (idx >> uint(j)) ^ 1
			authNode := FORSNode(hs, skSeed, pkSeed, i*(1<<uint(j))+s, j, adrs, p)
			sig = append(sig, authNode...)
		}
	}

	return sig
}

// FORSPKFromSig computes the FORS public key from a signature.
// FIPS 205 Algorithm 15: fors_pkFromSig(SIG_FORS, md, PK.seed, ADRS).
func FORSPKFromSig(hs HashSuite, sigFORS, md, pkSeed []byte, adrs *Address, p *Params) []byte {
	n := p.N
	k := p.K
	a := p.A

	indices := base2b(md, a, k)

	roots := make([]byte, 0, k*n)
	offset := 0

	for i := 0; i < k; i++ {
		idx := indices[i]

		// Leaf from secret key
		sk := sigFORS[offset : offset+n]
		offset += n

		adrs.SetType(AddrForsTree)
		adrs.SetTreeHeight(0)
		adrs.SetTreeIndex(uint32(i*(1<<uint(a)) + idx))
		node := hs.F(pkSeed, adrs, sk, n)

		// Walk up authentication path
		for j := 0; j < a; j++ {
			authNode := sigFORS[offset : offset+n]
			offset += n

			adrs.SetTreeHeight(uint32(j + 1))

			if (idx>>uint(j))%2 == 0 {
				adrs.SetTreeIndex(uint32(i*(1<<uint(j)) + (idx >> uint(j+1))))
				concat := make([]byte, 0, 2*n)
				concat = append(concat, node...)
				concat = append(concat, authNode...)
				node = hs.H(pkSeed, adrs, concat, n)
			} else {
				adrs.SetTreeIndex(uint32(i*(1<<uint(j)) + (idx >> uint(j+1))))
				concat := make([]byte, 0, 2*n)
				concat = append(concat, authNode...)
				concat = append(concat, node...)
				node = hs.H(pkSeed, adrs, concat, n)
			}
		}

		roots = append(roots, node...)
	}

	// Hash all roots together
	forsRootsAdrs := adrs.Copy()
	forsRootsAdrs.SetType(AddrForsRoots)
	forsRootsAdrs.SetKeyPairAddress(adrs.GetKeyPairAddress())
	return hs.Tl(pkSeed, &forsRootsAdrs, roots, n)
}

File: swift/Sources/PQCStandards/SlhDsa/Fors.swift
// SLH-DSA FORS (Forest of Random Subsets)

public enum Fors {
    /// FORS tree leaf
    public static func forsLeaf(params: SlhDsaParams, skSeed: [UInt8], pkSeed: [UInt8],
                                adrs: inout SlhAddress, idx: UInt32) -> [UInt8] {
        let kpAddr = adrs.getKeyPairAddress()
        adrs.setType(SlhAddress.forsKeyPrf)
        adrs.setKeyPairAddress(kpAddr)
        adrs.setTreeIndex(idx)
        let sk = SlhHash.prf(params: params, pkSeed: pkSeed, skSeed: skSeed, adrs: adrs)
        adrs.setType(SlhAddress.forsTree)
        adrs.setKeyPairAddress(kpAddr)
        adrs.setTreeHeight(0)
        adrs.setTreeIndex(idx)
        return SlhHash.f(params: params, pkSeed: pkSeed, adrs: adrs, m: sk)
    }

    /// FORS tree node
    public static func forsNode(params: SlhDsaParams, skSeed: [UInt8], pkSeed: [UInt8],
                                adrs: inout SlhAddress, idx: UInt32, height: Int) -> [UInt8] {
        if height == 0 {
            return forsLeaf(params: params, skSeed: skSeed, pkSeed: pkSeed, adrs: &adrs, idx: idx)
        }
        let left = forsNode(params: params, skSeed: skSeed, pkSeed: pkSeed,
                           adrs: &adrs, idx: 2 * idx, height: height - 1)
        let right = forsNode(params: params, skSeed: skSeed, pkSeed: pkSeed,
                            adrs: &adrs, idx: 2 * idx + 1, height: height - 1)
        let kpAddr = adrs.getKeyPairAddress()
        adrs.setType(SlhAddress.forsTree)
        adrs.setKeyPairAddress(kpAddr)
        adrs.setTreeHeight(UInt32(height))
        adrs.setTreeIndex(idx)
        return SlhHash.h(params: params, pkSeed: pkSeed, adrs: adrs, m1: left, m2: right)
    }

    /// FORS Sign
    public static func sign(params: SlhDsaParams, md: [UInt8], skSeed: [UInt8],
                           pkSeed: [UInt8], adrs: inout SlhAddress) -> [UInt8] {
        let indices = SlhHash.messageToIndices(md: md, k: params.k, a: params.a)
        let kpAddr = adrs.getKeyPairAddress()
        var sig = [UInt8]()

        for i in 0..<params.k {
            let idx = indices[i]
            let treeOffset = UInt32(i) * (1 << params.a)

            // Secret key value
            adrs.setType(SlhAddress.forsKeyPrf)
            adrs.setKeyPairAddress(kpAddr)
            adrs.setTreeIndex(treeOffset + idx)
            let sk = SlhHash.prf(params: params, pkSeed: pkSeed, skSeed: skSeed, adrs: adrs)
            sig.append(contentsOf: sk)

            // Auth path: at each level j, include the sibling subtree root
            for j in 0..<params.a {
                // The absolute index of the leaf's ancestor at height j
                let ancestorIdx = (treeOffset + idx) >> j
                // The sibling has absolute index: ancestor ^ 1
                let siblingIdx = ancestorIdx ^ 1
                let sibNode = forsNode(params: params, skSeed: skSeed, pkSeed: pkSeed,
                                      adrs: &adrs, idx: siblingIdx, height: j)
                sig.append(contentsOf: sibNode)
            }
        }
        return sig
    }

    /// FORS pk from sig
    public static func pkFromSig(params: SlhDsaParams, sig: [UInt8], md: [UInt8],
                                 pkSeed: [UInt8], adrs: inout SlhAddress) -> [UInt8] {
        let n = params.n
        let indices = SlhHash.messageToIndices(md: md, k: params.k, a: params.a)
        let kpAddr = adrs.getKeyPairAddress()
        var roots = [UInt8]()

        var sigOff = 0
        for i in 0..<params.k {
            let idx = indices[i]
            let sk = Array(sig[sigOff..<(sigOff + n)])
            sigOff += n

            // Compute leaf from sk
            adrs.setType(SlhAddress.forsTree)
            adrs.setKeyPairAddress(kpAddr)
            adrs.setTreeHeight(0)
            let treeIdx = UInt32(i) * (1 << params.a) + idx
            adrs.setTreeIndex(treeIdx)
            var node = SlhHash.f(params: params, pkSeed: pkSeed, adrs: adrs, m: sk)

            // Walk up the tree
            for j in 0..<params.a {
                let authNode = Array(sig[sigOff..<(sigOff + n)])
                sigOff += n

                adrs.setTreeHeight(UInt32(j + 1))
                let parentIdx = treeIdx >> (j + 1)
                adrs.setTreeIndex(parentIdx)

                if ((treeIdx >> j) & 1) == 0 {
                    node = SlhHash.h(params: params, pkSeed: pkSeed, adrs: adrs,
                                    m1: node, m2: authNode)
                } else {
                    node = SlhHash.h(params: params, pkSeed: pkSeed, adrs: adrs,
                                    m1: authNode, m2: node)
                }
            }
            roots.append(contentsOf: node)
        }

        adrs.setType(SlhAddress.forsRoots)
        adrs.setKeyPairAddress(kpAddr)
        adrs.setTreeHeight(0)
        adrs.setTreeIndex(0)
        return SlhHash.tl(params: params, pkSeed: pkSeed, adrs: adrs, m: roots)
    }
}
